pyomo/contrib/solver/gurobi_direct.py

```diff
@@ -0,0 +1,419 @@
+                # gurobi_model.update()
+                timer.stop('transfer_model')
+
+                options = config.solver_options
+
+                gurobi_model.setParam('LogToConsole', 1)
+
+                if config.threads is not None:
+                    gurobi_model.setParam('Threads', config.threads)
+                if config.time_limit is not None:
+                    gurobi_model.setParam('TimeLimit', config.time_limit)
+                if config.rel_gap is not None:
+                    gurobi_model.setParam('MIPGap', config.rel_gap)
+                if config.abs_gap is not None:
+                    gurobi_model.setParam('MIPGapAbs', config.abs_gap)
+
+                if config.use_mipstart:
+                    raise MouseTrap("MIPSTART not yet supported")
+
+                for key, option in options.items():
+                    gurobi_model.setParam(key, option)
+
+                timer.start('optimize')
+                gurobi_model.optimize()
+                timer.stop('optimize')
+        finally:
+            os.chdir(orig_cwd)
+
+        res = self._postsolve(
+            timer,
+            config,
+            GurobiDirectSolutionLoader(
+                gurobi_model, A, x, repn.rows, repn.columns, repn.objectives
+            ),
+        )
+        res.solver_configuration = config
+        res.solver_name = 'Gurobi'
+        res.solver_version = self.version()
+        res.solver_log = ostreams[0].getvalue()
+
+        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        res.timing_info.start_timestamp = start_timestamp
+        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.timer = timer
+        return res
+
+    def _postsolve(self, timer: HierarchicalTimer, config, loader):
+        grb_model = loader._grb_model
+        status = grb_model.Status
+
+        results = Results()
+        results.solution_loader = loader
+        results.timing_info.gurobi_time = grb_model.Runtime
+
+        if grb_model.SolCount > 0:
+            if status == gurobipy.GRB.OPTIMAL:
+                results.solution_status = SolutionStatus.optimal
+            else:
+                results.solution_status = SolutionStatus.feasible
+        else:
+            results.solution_status = SolutionStatus.noSolution
+
+        results.termination_condition = self._get_tc_map().get(
+            status, TerminationCondition.unknown
+        )
+
+        if (
+            results.termination_condition
+            != TerminationCondition.convergenceCriteriaSatisfied
+            and config.raise_exception_on_nonoptimal_result
+        ):
+            raise RuntimeError(
+                'Solver did not find the optimal solution. Set '
+                'opt.config.raise_exception_on_nonoptimal_result=False '
+                'to bypass this error.'
+            )
+
+        if loader._pyo_obj:
```

Comment:
Again, not necessarily related to this PR, but I look at this line and question it. What is `_pyo_obj` and why is it relevant?

Reply:
That is the Pyomo Objective object that was emitted to the solver (if the Pyomo model has an objective).  We use it to control if we ask Gurobi for the model objective value (it is legal to give Gurobi a feasibility problem with no objective, but if you do, it is an error to ask for the final objective value).